### PR TITLE
ENH: Install FSL and ANTs within GitHub Actions

### PR DESCRIPTION
**Pull request by josephmje**
Replaces #173

### CODE REVIEW

**Comment by josephmje** [comment]
There are usually some extra FSL environment variables and I'm not sure which ones are necessary yet:
```
PATH: /opt/fsl/bin:$PATH
FSLOUTPUTTYPE: NIFTI_GZ
FSLMULTIFILEQUIT: true
FSLTCLSH: /opt/fsl/bin/fsltclsh
FSLWISH: /opt/fsl/bin/fslwish
LD_LIBRARY_PATH: /opt/fsl:$LD_LIBRARY_PATH
```

**Review by jhlegarreta — COMMENTED** [review]
Thanks for doing this :+1: :100:.

In order to check that this is working, the preprocessing solutions notebook (the non-solution notebook has in-line exercises) should be added to the list tested notebooks.

I'd have voted for separating the changes related to #172; in this PR the set of related changes is the same, and if the FSL/ANTs installation requires further changes, the previous changes will not get in. Anyways, we can see if the builds pass when the preprocessing notebook is added, and if it does, we can dismiss #172.

**Comment by josephmje** [comment]
> I'd have voted for separating the changes related to #172 

My mistake. I had to include those commands in order for the tests to run.

I will add the preprocessing notebooks to see if everything works fine.

**Review by jhlegarreta — COMMENTED** [review]
Thanks for the effort @josephmje. It is looking good. An inline comment.

The build failures are due to the inaccurate data folder path. I will submit a separate PR.

@kaitj Cross-referencing https://github.com/carpentries-incubator/SDC-BIDS-dMRI/pull/175#pullrequestreview-722573419. I think it is best separating this from the changes in #172. Otherwise, as you see here, there are no differences in the related parts. So if you think #172 can be improved, proposing the changes there and getting them merged if they work would prevent from merging here a part that may otherwise go unnoticed. Unless changing and testing your thought will take a bit of time. Thanks.

**Comment by kaitj** [comment]
> @kaitj Cross-referencing #175 (review). I think it is best separating this from the changes in #172. Otherwise, as you see here, there are no differences in the related parts. So if you think #172 can be improved, proposing the changes there and getting them merged if they work would prevent from merging here a part that may otherwise go unnoticed. Unless changing and testing your thought will take a bit of time. Thanks.

I think whenever these PRs are ready, you can go ahead and merge them in. If this is referring to setting up the triggers for push vs pull request, I would need to do some testing first and can submit as a separate PR. I haven't had a chance to do that yet (tied up with some other commitments this week).

**Comment by jhlegarreta** [comment]
> I think whenever the current PRs are to everyone's satisfaction, you can go ahead and merge them in. If this is referring to setting up the triggers for push vs pull request, I would need to do some testing first and can submit as a separate PR. I haven't had a chance to do that yet (tied up with some other commitments this week).

OK, so let's merge #172 and refinements or fixes if necessary will be proposed as separate PRs.

File: .github/workflows/build-test.yaml [hunk 1]
```diff
@@ -1,39 +1,101 @@
...
+    # Run with multiple Python versions
     - name: Setup Python ${{ matrix.python-version }}
       uses: actions/setup-python@v2
       with:
         python-version: ${{ matrix.python-version }}
 
-    # Use a cache for the pip install to avoid reinstalling every time this is run.
-    - name: Cache pip install
-      uses: actions/cache@v2
+    # Install Python requirements
```
[the remark this answers is not in the document]
Comment (josephmje) [reply .github/workflows/build-test.yaml:90]: Good point. I'll add in a name for the rest of the caches.

**Comment by jhlegarreta** [comment]
Thanks for all this effort @josephmje. Not sure why tests are reported as successful since the detailed report for the preprocessing notebook contains failures:
```
============= 5 passed, 1 xfailed, 5 xpassed in 2272.58s (0:37:52) =============
```
https://github.com/carpentries-incubator/SDC-BIDS-dMRI/runs/3254299943?check_suite_focus=true#step:13:79

But for the sake of get things going, and given that this is already a big leap forward because it succeeds in installing the necessary dependencies for the notebook to run, which was the primary aim, I'd be for investigating the reported issues in separate PRs.

**Comment by josephmje** [comment]
> But for the sake of get things going, and given that this is already a big leap forward because it succeeds in installing the necessary dependencies for the notebook to run, which was the primary aim, I'd be for investigating the reported issues in separate PRs, unless fixing them is easy.

Thank you @jhlegarreta! I think I'm done all I can for this PR. I at least wanted to make sure that the FSL commands were findable.

**Comment by josephmje** [comment]
It looks like the FSL cache and the apt-get cache are failing to be saved.

```
Post job cleanup.
/usr/bin/tar --posix --use-compress-program zstd -T0 -cf cache.tzst -P -C /home/runner/work/SDC-BIDS-dMRI/SDC-BIDS-dMRI --files-from manifest.txt
/usr/bin/tar: ../../../../../opt/fsl/fslpython/pkgs/conda-package-handling-1.7.3-py39h27cfd23_1_i3f7jwg: Cannot open: Permission denied
/usr/bin/tar: ../../../../../opt/fsl/fslpython/pkgs/yaml-0.2.5-h7b6447c_08j1ypz_r: Cannot open: Permission denied
/usr/bin/tar: ../../../../../opt/fsl/fslpython/pkgs/zlib-1.2.11-h7b6447c_3afzd2vr9: Cannot open: Permission denied
/usr/bin/tar: ../../../../../opt/fsl/fslpython/pkgs/wxpython-4.0.6-py37h10a64c0_1/info/licenses/LICENSE.txt: Cannot open: Permission denied
/usr/bin/tar: ../../../../../opt/fsl/etc/xtract_data/Macaque/ac/exclude.nii.gz: Cannot open: Permission denied
/usr/bin/tar: Exiting with failure status due to previous errors
Warning: Tar failed with error: The process '/usr/bin/tar' failed with exit code 2
```

```
 Post job cleanup.
/usr/bin/tar --posix --use-compress-program zstd -T0 -cf cache.tzst -P -C /home/runner/work/SDC-BIDS-dMRI/SDC-BIDS-dMRI --files-from manifest.txt
/usr/bin/tar: ../../../../../var/lib/apt/lists/lock: Cannot open: Permission denied
/usr/bin/tar: ../../../../../var/lib/apt/lists/partial: Cannot open: Permission denied
/usr/bin/tar: Exiting with failure status due to previous errors
Warning: Tar failed with error: The process '/usr/bin/tar' failed with exit code 2
```

**Comment by jhlegarreta** [comment]
> It looks like the FSL cache and the apt-get cache are failing to be saved.

I don't have a good suggestion for this. Not sure if it is a permission issue (and if yes, how these are expected to be changed), or whether it has to do more with an issue with the path.

**Review by jhlegarreta — APPROVED** [review]

**Comment by kaitj** [comment]
Do we know where the cache file is getting stored (eg. what directory that `cache.tzst` file is)? Is it possible its trying to save it in a location where `sudo` is needed? 

The second snippet of code @josephmje provided seems to be an issue others have encountered too - https://github.com/actions/cache/issues/324

**Comment by josephmje** [comment]
> Do we know where the cache file is getting stored (eg. what directory that `cache.tzst` file is)? Is it possible its trying to save it in a location where `sudo` is needed?

It could be worth trying to install ANTs and FSL in the home folder maybe?

> The second snippet of code @josephmje provided seems to be an issue others have encountered too - [actions/cache#324](https://github.com/actions/cache/issues/324)

Thanks for finding this @kaitj. Maybe it isn't worthwhile trying to cache the apt-get step since it is pretty quick anyways.

**Comment by jhlegarreta** [comment]
For the sake of making things move forward, last reported persistent issues (`XFAIL`/`XPASS` in preprocessing notebook and post job cleanups) may be best addressed separately.

**Comment by josephmje** [comment]
> For the sake of making things move forward, last reported persistent issues (`XFAIL`/`XPASS` in preprocessing notebook and post job cleanups) may be best addressed separately.

I'm in favour of this. Will merge now

**Pull request merged**